packages/flutter/lib/src/cupertino/colors.dart

```diff
@@ -80,3 +93,608 @@ class CupertinoColors {
+
+  /// The color to use when the [BuildContext] implies a combination of dark mode,
+  /// normal contrast, and elevated interface elevation.
+  ///
+  /// Defaults to [defaultColor] when unspecified.
+  Color get darkElevatedColor => _colorMap[5];
+
+  /// The color to use when the [BuildContext] implies a combination of light mode,
+  /// high contrast, and elevated interface elevation.
+  ///
+  /// Defaults to [defaultColor] when unspecified.
+  Color get highContrastElevatedColor => _colorMap[6];
+
+  /// The color to use when the [BuildContext] implies a combination of dark mode,
+  /// high contrast, and elevated interface elevation.
+  ///
+  /// Defaults to [defaultColor] when unspecified.
+  Color get darkHighContrastElevatedColor => _colorMap[7];
+
+  final List<Color> _colorMap;
+
+  /// Resolves the given [Color] to a concrete [Color], using the given [BuildContext].
+  /// If the given color is already a concrete [Color], it will be returned as is.
+  static Color resolve(Color resolvable, BuildContext context, { bool nullOk = false })
+    => (resolvable is CupertinoDynamicColor)
+      ? resolvable.resolveFrom(context, nullOk: nullOk)
+      : resolvable;
+
+  /// Resolves this `CupertinoDynamicColor` using the provided [BuildContext].
+  ///
+  /// Calling this method will create a new `CupertinoDynamicColor` that is almost
+  /// identical to the orignal `CupertinoDynamicColor` except with a different
+  /// [defaultColor], such that when the new `CupertinoDynamicColor` is used as
+  /// a regular [Color], it is equivalent to one of the specified colors that
+  /// matches the given [BuildContext]'s traits.
+  ///
+  /// For example, if the given [BuildContext] indicates the widgets in the subtree
+  /// should be displayed in dark mode, with high accessibility contrast and an
+  /// elevated interface elevation, the resolved `CupertinoDynamicColor` will be
+  /// the same as this [CupertinoDynamicColor], except that its [defaultColor]
+  /// is `darkHighContrastElevatedColor` from the orignal `CupertinoDynamicColor`.
+  ///
+  /// Calling this function may create dependencies on the closest instance of some
+  /// [InheritedWidget]s that enclose the given [BuildContext]. E.g., if [darkColor]
+  /// is different from [normalColor], this function will call [CupertinoTheme.of],
+  /// and then [MediaQuery.of] if brightness wasn't specified in the theme data
+  /// retrived from the previous [CupertinoTheme.of] call, in an effort to determine
+  /// the brightness value.
+  ///
+  /// If any of the required dependecies is missing from the given context, an exception
+  /// will be thrown unless [nullOk] is set to `true`.
+  CupertinoDynamicColor resolveFrom(BuildContext context, { bool nullOk = false }) {
+    int dependencyBitMask = 0, configBitMask = 0;
+    final List<Color> colors = _colorMap;
+
+    for (int i = 0; i < colors.length; i++) {
+      for(int bitShift = 0; bitShift < 3; bitShift ++) {
```

Comment:
Note sure why we're doing this if all 8 colors are always defined? Could we just define private getters like 
```
bool get _dependsOnPlatformBrightness {
  return normalColor != darkColor || elevatedColor != darkElevatedColor || highContrastElevatedColor != darkHighContrastElevatedColor;
}
```